### PR TITLE
Refactoring debts

### PR DESCRIPTION
**Pull request by marcosvrs**
* Apply all Strict Type-Checking options from TS;
* Apply all Additional Checks options from TS;
* Remove all overwritten rules from tslint (except no-relative-imports);
* Fix all erros;
* Use namespaces for Factories;
* Rely on interfaces;
* Add Jest for a faster test run;
* Fix Travis for getting the latest tplant npm version.
* Add new tests for namespace generation;

### CODE REVIEW

**Comment by marcosvrs** [comment]
@bafolts I tried to find a configuration for our webpack to allow also usage of our code as a module. But I fail.
And after studying other repositories like [TypeScript](https://github.com/Microsoft/TypeScript) and [tslint](https://github.com/palantir/tslint), I realize that none of them uses bundler like webpack. At most, they use typescript togheter with gulp for compilation. And leave all the files separated at the end of the build.
Which, now I believe, is the right way.

**Comment by bafolts** [comment]
That makes sense. A tool like webpack makes sense more if packaging for the web.

File: src/Components/Namespace.ts [hunk 1]
```diff
@@ -23,8 +23,10 @@ export class Namespace implements IComponentComposite {
         }
         result.push(declaration.join(''));
         this.parts.forEach((part: IComponentComposite): void => {
-            result.push(part.toPUML()
-            .replace(/^/gm, '    '));
+            result.push(
+                part.toPUML()
+                    .replace(/^(?!\s*$)/gm, '    ')
```
Comment (bafolts) [inline src/Components/Namespace.ts]: What is this regular expression attempting to replace?

**Comment by bafolts** [comment]
This still looks fine to merge to me.

File: src/Components/Namespace.ts (hunk 1 above)
Comment (marcosvrs) [reply src/Components/Namespace.ts]: Starting of line except empty lines. To add indentation.

**Pull request merged**